### DISCUSSION TITLE
Overall time restriction

### DISCUSSION THREAD

**Discussion by einGlasRotwein** [jspsych/jsPsych#1229] [Q&A] [answered]
Hello everyone,

I created an experiment where the duration is not determined by the number of trials, but rather, the experiment should stop after a certain time (say, 5 minutes). I was able to implement this as a conditional timeline, i.e. for every trial, I created an if node like in [this example](https://www.jspsych.org/overview/timeline/#conditional-timelines) and then chained those if nodes in a procedure. Here is a simplified example, with an pverall time limit of 10 seconds. Depending on how long you take to answer the `hello_world` trials, you will reach the time limit or not:

```js
// TRIALS
var fixation = {
    type: 'html-keyboard-response',
    stimulus: '+',
    choices: jsPsych.NO_KEYS,
    trial_duration: jsPsych.timelineVariable('fixation_duration'),
}

var if_fixation = {
    timeline: [fixation],
    conditional_function: function(){
        // get the data from the previous trial,
        // and check how much time has passed until the end of the last trial
        var data = jsPsych.data.get().last(1).values()[0];
        if(data.time_elapsed > 10000){
            return false;
        } else {
            return true;
        }
    }
}

var hello_world = {
    type: 'html-keyboard-response',
    stimulus: 'hello world'
}

var if_hello_world = {
    timeline: [hello_world],
    conditional_function: function(){
        // get the data from the previous trial,
        // and check how much time has passed until the end of the last trial
        var data = jsPsych.data.get().last(1).values()[0];
        if(data.time_elapsed > 10000){
            return false;
        } else {
            return true;
        }
    }
}

// TIMELINE
var timeline = [];

// Put various fixation durations into an array
var fixation_durations = [];

var durations = [800, 1600, 1000, 800];

for (let i = 0; i < durations.length; i++) {
    fixation_durations.push({fixation_duration: durations[i]});
}

// Set up a procedure
var PVT_procedure = {
    timeline: [if_fixation, if_hello_world],
    timeline_variables: fixation_durations
}

timeline.push(PVT_procedure);
```

This works, but is this "good practice"? Are there more elegant solutions for an overall timelimit?

Thank you in advance!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @einGlasRotwein ,

We don't have a particularly elegant way to handle this situation yet. One other option could be to use the `on_trial_finish` event handler in `jsPsych.init()`. This can set up a function to run after every trial. You could compare to some clock here, and if the maximum time has been exceeded you can call `jsPsych.endExperiment()`. That said, I kind of prefer your version because it gives more control over when the experiment can end.

**Reply by einGlasRotwein** [reply]
Thank you, I will leave my code as it is then :-)

**Comment by sankalp35** [comment]
Thanks for this @einGlasRotwein, it's rather helpful. I have been struggling to put a time limit on my task -- one would have hoped that would be rather easy. Your solution is helpful!
